### PR TITLE
use fuzzywuzzymit

### PR DESCRIPTION
**Pull request by graingert**

### CODE REVIEW

**Comment by cgivre** [comment]
@graingert Thank you!! This is awesome!  I have a suite of string matching functions which I'd like to submit to Apache Drill, some of which wrap fuzzywuzzy functionality, so now that you've made this available, I'll submit a PR to Drill!

**Comment by josegonzalez** [comment]
Hope exactly is your version mit licensed? It seems as though you just copy pasted fuzzywuzzy and changed an import.

I am not a lawyer, but it was explained to me that such move would not be sufficient. If it was, why would that require a subpackage?

**Comment by graingert** [comment]
> changed an import.

Specifically I removed an import of GPL code. fuzzywuzzy itself is MIT
licensed. (Look at the headers on fuzz.py etc)

**Comment by josegonzalez** [comment]
Right. Couldn't that have been some in this to if that were a valid way to relicense?

**Comment by graingert** [comment]
What?

**Comment by josegonzalez** [comment]
My understanding of the GPL is that once you pull in any GPL code, the entire codebase gets licensed as GPL. Simply removing the dependency from the codebase does not remove the requirement of licensing the code as GPL - that's the viral bit of the license.

What you've done is just the above. You forked the fuzzywuzzy codebase and changed the import of SequenceMatcher to pull from an external library. Again, I am not a lawyer, but I don't believe that allows you to relicense fuzzywuzzy as MIT.

If that is indeed the case - and if it is, I'd love to see references so I can respond to the occasional email I get from folks about this exact project - then why not submit that single change to this repository?

In short, I don't see any benefits from the indirection here, and only see a further GPL violation.

Please feel free to correct me if I'm wrong - with sources please! - as I personally dislike the GPL and it is very frustrating to me that we have to license it this way in the first place.

**Comment by graingert** [comment]
> My understanding of the GPL is that once you pull in any GPL code, the entire codebase gets licensed as GPL. Simply removing the dependency from the codebase does not remove the requirement of licensing the code as GPL - that's the viral bit of the license.

This is not the case: all that happens is your right to use, link to and distribute python-Levenshtein program is revoked until you come into compliance.

"To come into compliance" you must either cease to use, link to and or distribute the python-Levenshtein or provide your source code under the GPL license, with no extra restrictions. You as copyright holder can provide fuzzywuzzy source code under other compatible licenses if you wish. (eg you could make your part MIT, but you could not make it proprietary)

**Comment by graingert** [comment]
> then why not submit that single change to this repository?

because `fuzzywuzzy` is faster than `fuzzywuzzymit`

**Comment by graingert** [comment]
with references to the GPLv2 itself: https://www.gnu.org/licenses/old-licenses/gpl-2.0.en.html

>      b) You must cause any work that you distribute or publish, that in whole or in part contains or is derived from the Program or any part thereof, to be licensed as a whole at no charge to all third parties under the terms of this License. 

Note that it does not say "You must cause any work that you distribute or publish ... to *only* be licensed ... under the terms of this License"

>  Thus, it is not the intent of this section to claim rights or contest your rights to work written entirely by you; rather, the intent is to exercise the right to control the distribution of derivative or collective works based on the Program. 

When you no longer link to, use or distribute the Program, python-Levenshtein loses all rights to control the distribution of your Program.

**Comment by josegonzalez** [comment]
Again, I don't think you can relicense your fork as MIT. MIT => GPLv2 is okay, but not in the reverse.

[Here is an issue](https://github.com/seatgeek/fuzzywuzzy/issues/130) where we discussed trying to include _just_ the individual license on the StringMatcher library.

Even if it were possible for me to relicense it, there are a couple issues here:

- You are not the copyright holder of fuzzywuzzy, SeatGeek is. According to the GPLv2, you cannot relicense your fork. If you made a clean-room implementation, its potentially possible, but see [Google v Oracle](https://www.theinquirer.net/inquirer/news/3029302/a-google-vs-oracle-copyright-ruling-could-well-kill-small-scale-open-source-projects) for a ruling where we might be out of luck.
- We cannot relicense from GPLv2 to MIT without getting express permission from everyone who has contributed for the lifetime of the project. Twitter Bootstrap [did this](https://github.com/twbs/bootstrap/issues/2054) and it took forever. We could probably do this, but maintaining this library from an OSS perspective isn't strictly part of my work responsibilities, and we don't have the resources to do so.

My understanding here is that only SeatGeek - as the copyright holder - can relicense the project from GPLv2 to MIT, and only once we stop using python-Levenshtein completely, and then _only with_ express permission of _all_ those who have contributed to the project. Others are free to fork our project, but GPLv2 requires that they keep the same license or use a compatible license (and only for new changes).

[Copyleft licenses (GPLv2) are not compatible with Permissive licenses (MIT)](https://blog.gerv.net/2013/02/relicensing-when-do-you-have-to-ask/), meaning no one is currently allowed to relicense fuzzywuzzy or derivatives to MIT.

> Note: I am not a lawyer, I am merely referencing other reading online.

**Comment by graingert** [comment]
@josegonzalez I'm not relicensing it: https://github.com/graingert/fuzzywuzzymit/blob/master/fuzzywuzzymit/fuzz.py#L6

**Comment by josegonzalez** [comment]
The license on that file in our repository is actually wrong, the entire project is GPLv2 (yay viral licenses). I'll remove that from our codebase.

As your project is a derivative of SeatGeek's fuzzywuzzy - I hope thats not up to debate, as you've actually forked the project, which makes that change clear - your project is _also_ GPLv2 licensed. Therefore, you are relicensing the project.

**Comment by graingert** [comment]
@josegonzalez it's not wrong. it's perfectly valid to release under both MIT and GPL

**Comment by josegonzalez** [comment]
It is indeed valid to release under both, but only under the following conditions:

- A project was previously - and correctly! - released under MIT. fuzzywuzzy never was, as we included `StringMatcher.py` in the source out of the gate.
- Derivatives must use a compatible license. MIT is not a compatible license with GPLv2, at least in the way you are relicensing.
- Only the copyright holder - SeatGeek - can relicense. You are not the copyright holder, so you cannot relicense.
- The copyright holder must get permission from all contributors in order to relicense. We have the ability to relicense all contributions from SeatGeek employees, but would have to reach out to all other contributors and confirm that their changes are okay to be relicensed.

**Comment by graingert** [comment]
> It is indeed valid to release under both, but only under the following conditions:

>   *  A project was previously - and correctly! - released under MIT. fuzzywuzzy never was, as we included StringMatcher.py in the source out of the gate.
You licensed it under MIT but were violating the GPL. This doesn't make your code licensed under the GPL. It should have been, but you didn't
>    * Derivatives must use a compatible license. MIT is not a compatible license with GPLv2, at least in the way you are relicensing.
MIT is compatible with the GPL. I'm also not relicensing, I'm just using the MIT license as stated in the files
>    * Only the copyright holder - SeatGeek - can relicense. You are not the copyright holder, so you cannot relicense.
True, I'm not relicensing though.
>    * The copyright holder must get permission from all contributors in order to relicense. We have the ability to relicense all contributions from SeatGeek employees, but would have to reach out to all other contributors and confirm that their changes are okay to be relicensed.
False. The copyright holder has the exclusive right to relicense. However in some cases the contributors are the copyright holder. In this case the copyright holder is Adam Cohen.

**Comment by graingert** [comment]
Currently fuzzywuzzy is still under the MIT only, because on no files have you included the GPL notice: https://github.com/seatgeek/fuzzywuzzy/blob/master/LICENSE.txt#L289-L309

**Comment by josegonzalez** [comment]
> Note: Adam is one of my coworkers, if we can nail down the specifics here, I'm like 100% sure he'd be fine with giving the ok here.

I don't want to argue about this. Neither of us are lawyers here. Do you know of any lawyer who is aware of OSS licensing that can weigh in here? I've been asking in some online chat rooms and the answers are... mixed.

**Comment by graingert** [comment]
@josegonzalez now fuzzywuzzy has no license. It just includes an inert copy of the MIT license, you need to follow these instructions to apply it:

https://github.com/seatgeek/fuzzywuzzy/blob/master/LICENSE.txt#L289-L309

**Comment by josegonzalez** [comment]
I'm going to close this. We are currently not interested in relicensing the project, and this change makes it more difficult for those using fuzzywuzzy to figure out where they should contribute changes.

If any user wishes to use a potentially MIT licensed version of the project, feel free to use the fork. If someone has access to a copyright lawyer that would like to spend some time figuring out whether there is a valid path to licensing the codebase under the MIT license, feel free to put them in touch with.

Apologies @graingert for the back and forth, and thank you for taking the time to explain licensing issues to me.
